jxl/src/frame/modular/transforms/squeeze.rs

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+use crate::{
+    error::{Error, Result},
+    frame::modular::{ChannelInfo, ModularBufferInfo},
+    headers::modular::SqueezeParams,
+};
+
+use super::TransformStep;
+
+use crate::util::tracing_wrappers::*;
+
+pub fn do_squeeze_step(
+    _step: TransformStep,
+    _buffers: &mut [ModularBufferInfo],
+    (_gx, _gy): (usize, usize),
+) -> Result<Vec<(usize, usize)>> {
+    todo!()
+}
+
+#[instrument(level = "trace", err)]
+pub fn check_squeeze_params(
+    channels: &[(usize, ChannelInfo)],
+    params: &SqueezeParams,
+) -> Result<()> {
+    let end_channel = (params.begin_channel + params.num_channels) as usize;
+    if end_channel > channels.len() {
+        return Err(Error::InvalidChannelRange(
+            params.begin_channel as usize,
+            params.num_channels as usize,
+            channels.len(),
+        ));
+    }
+    if channels[params.begin_channel as usize].1.is_meta() != channels[end_channel - 1].1.is_meta()
+    {
+        return Err(Error::MixingDifferentChannels);
+    }
+    if channels[params.begin_channel as usize].1.is_meta() && !params.in_place {
+        return Err(Error::MetaSqueezeRequiresInPlace);
+    }
+    Ok(())
+}
+
+pub fn default_squeeze(data_channel_info: &[(usize, ChannelInfo)]) -> Vec<SqueezeParams> {
+    let mut w = data_channel_info[0].1.size.0;
+    let mut h = data_channel_info[0].1.size.1;
+    let nc = data_channel_info.len();
+
+    let mut params = vec![];
+
+    let num_meta_channels = data_channel_info
+        .iter()
+        .take_while(|x| x.1.is_meta())
+        .count();
+
+    if nc > 2 && data_channel_info[1].1.size == (w, h) {
+        // 420 previews
+        let sp = SqueezeParams {
+            horizontal: true,
+            in_place: false,
+            begin_channel: num_meta_channels as u32 + 1,
+            num_channels: 2,
+        };
+        params.push(sp);
+        params.push(SqueezeParams {
+            horizontal: false,
+            ..sp
+        });
+    }
+
+    const MAX_FIRST_PREVIEW_SIZE: usize = 8;
+
+    let sp = SqueezeParams {
+        begin_channel: num_meta_channels as u32,
+        num_channels: nc as u32,
```

Comment:
should this be `nc - num_meta_channels` if it only starts at ben_channel?

Reply:
(I only moved this code)
I *think* this was correct, I think it matches the C++ code and Jon checked it, but of course my confidence level is low...